### DISCUSSION TITLE
negative values in temporal components

### DISCUSSION THREAD

**Discussion by wienecke** [flatironinstitute/CaImAn#1252] [Q&A]
Hi 
I use CNMF.fit, then CNMF.refit, to extract components from 2-photon GCaMP7f movies. 
I use method_init graph_nmf. I skip deconvolution by setting p to 0. I use patches because SNR varies across the FOV.

Below is a temporal component (estimates.C) plotted with estimates.view_components. 
It shows visual stimulus-evoked responses from a neuron. It is typical of my extracted traces. 
The “inferred trace” (red trace), is estimates.C
The “filtered raw data” (blue trace) is estimates.C + residuals
The positive parts of the blue trace are obscured by the red trace.

A linear function of the visual stimulus predicts the “filtered raw data” very well, but not the “inferred trace”
I expect this linear function to predict this neuron’s responses, so I’ve been using this model as a positive control of my source extractions as I adjust CaImAn input parameters.  
It’s clear from looking at the predictions that the model of the “filtered raw data” is better because it includes negative values.
Because of this, I don’t think these negative values are noise.
Based on my modeling, and my expectations, these negative values contain a similar amount of visually-evoked “signal” as the positive values.

But I thought that negative values in estimates.C represent noise.

Currently, in all my analysis, I use estimates.F_dff computed with use_residuals set to true. This gives me normalized traces with the negative parts preserved (i.e., it gives me something similar to a scaled version of the blue trace). And this gives me data that makes sense.

Can you help me understand these negative values, or why my “inferred trace” seems to be throwing away so much “signal” below zero? 
Might my CaImAn input parameters be adjusted to make the “inferred trace” include this “signal”?

Thank you!
Carl 

<img width="1495" alt="trace" src="https://github.com/flatironinstitute/CaImAn/assets/20496060/52b17b77-ee1a-4099-ad58-0bea98011b97">

**Comment by EricThomson** [comment]
Are you imaging dendrites or other non-somatic shapes? The `graph_nmf` init is not super-well vetted. I hope it works ok 😄 

I would be wary of skipping deconvolution: even if you don't care about spike counts, it can help you get better estimates of C during the source extraction step especially for lower SNR traces (this was one of the points of the original paper on CNMF: simultaneous denoising, demixing, deconvolution 😃 ): https://pubmed.ncbi.nlm.nih.gov/26774160/. 

Negative values can be very hard to interpret.  Have you tried toggling the `bas_nonneg` parameter?

**Comment by wienecke** [comment]
Thanks for your reply. Sorry for my slow reply; I wanted to test a few things first.

I am imaging non-somatic shapes, and I skip deconvolution by setting p=0 because my neurons are non-spiking.

I see the above problem with method_init 'greedy_roi', and 'sparse_nmf'. 

If I use deconvolution by setting p=1 or p=2, and I set bas_nonneg=False, then I do not see the above problem. So thank you for that suggestion. But since my neurons are nonspiking, I do set p=0, and bas_nonneg seems to have no effect if p=0.

My input movie is nonnegative, and it looks like negative values in the temporal component are introduced in line 214 of temporal.py, in function 'update_temporal_components', in this line: 
     YrA = YA - AA.T.dot(Cin).T 
And later, in line 124 of deconvolution.py, in function 'constrained_foopsi', those negatives are clipped to zero in this line:
     c = np.maximum(fluor, 0)

Can you help me understand the purpose of these lines (the first one, especially)? What is the meaning of the negative values after the first line? I don't understand why so much of the temporal component is getting clipped at the bottom. As I describe above, the clipped signal does not seem to be noise. 

Is it possible to make bas_nonneg=False have an effect when p=0? Or will that cause problems?

Thank you!
Carl

**Comment by kushalkolar** [comment]
It's pretty hard to say without seeing what the movie looks like, as well as what the reconstructed movie (AC) background (bf) and residuals (Y - AC - bf) look like.

**Comment by wienecke** [comment]
Thanks for your reply. 

I've attached the movie output by this call to play_movie:

   estimates.play_movie(images_sliced, q_min=1, q_max=99.75, gain_res=2, magnification=2, 
                                 include_bck=False, frame_range=slice(0,3048,1), bpx=False, thr=1, 
                                 save_movie=True, movie_name=pth_results, display=True, opencv_codec='H264',
                                 use_color=False, gain_color=4, gain_bck=0.2)

I've also attached the raw movie (imgs), reconstructed movie (Y_rec), background (B), and residuals (Y_res), which I saved from within play_movie. 

I've also attached my input parameters (pars).

Finally, I've attached the output of view_components, for one ROI (estimates.C at bottom). Like the example above, it shows clipping of low amplitude signal that correlates with an external variable roughly just as well as the parts that don't get clipped. I’ve also attached a scatterplot of that external variable vs. the same timeseries, and in red, the curve we expect to fit the scatterplot. You can see the lower amplitude part of the timeseries closely follows that curve. I’ve seen this in many ROIs of this cell type, so I don’t think it’s actually noise getting clipped. 

What do you think? Is a spatially and/or temporally distinct signal just being separated from a signal that better matches my input parameters? Could bas_nonneg=False fix this when p=0 (if it had an effect when p=0)?

Thank you!

Carl

[B.npy.zip](https://github.com/user-attachments/files/15812747/B.npy.zip)
[imgs.npy.zip](https://github.com/user-attachments/files/15812749/imgs.npy.zip)
[Y_rec.npy.zip](https://github.com/user-attachments/files/15812751/Y_rec.npy.zip)
[Y_res.npy.zip](https://github.com/user-attachments/files/15812753/Y_res.npy.zip)
[pars.npy.zip](https://github.com/user-attachments/files/15812889/pars.npy.zip)
<img width="991" alt="Screen Shot 2024-06-12 at 4 09 41 PM" src="https://github.com/flatironinstitute/CaImAn/assets/20496060/6a5e2a30-7b0a-4d59-9a38-a392786d979b">
<img width="533" alt="Screen Shot 2024-06-12 at 9 00 32 PM" src="https://github.com/flatironinstitute/CaImAn/assets/20496060/b687f5ab-1bb0-4af9-8fe9-3b400d8d3836">
https://github.com/flatironinstitute/CaImAn/assets/20496060/76edc909-e87a-4ca9-ac37-158bc005c04a

**Comment by kushalkolar** [comment]
Hi, I honestly don't have the bandwidth to go through your files, can you post a video of the input movie, and reconstructed movie, background and residuals. You can drag and drop vids upto 10MB on GitHub

**Comment by wienecke** [comment]
Hi
I've attached the movie you asked for below. From left to right, it shows: input movie, reconstructed movie, residuals, and background. It is a 727 KB .mov file.
Thanks!
Carl
https://github.com/flatironinstitute/CaImAn/assets/20496060/f70b688f-1d05-4a1f-b633-142f3cc7e7cd

**Comment by kushalkolar** [comment]
What are you imaging, this looks very different for what CNMF can do and seems to have high rank background. Have you tried CNMFE?

**Reply by wienecke** [reply]
Hi. What do you think of my reply below? Do you still recommend CNMFE after reading my reply? Why does bas_nonneg=False have no effect when p=0? Is this an 'issue'? Should I post this to the 'issues' thread? Thanks

**Comment by wienecke** [comment]
I'm 2-photon imaging GCaMP7f in axon terminals in the Drosophila brain at 5-8 Hz. The data above has been denoised with DeepCAD. But I see the same thing without denoising. Here, attached, is the same recording without the denoising. Also, I've included an example output of view_components, for one ROI in this recording without denoising, so you can see what's getting clipped in the timeseries when I don't include residuals. Finally, I've attached a screenshot of what happens when I include residuals (no clipping, on top), and do not include residuals (clipping, on bottom), for 3 ROIs from this recording without denoising. The red fit curves on top match how we expect activity in these ROIs to correlate with our external variable. The fits for the same ROIs on bottom do not match what we expect. It appears that valuable signal is lost to the residuals, with or without DeepCAD denoising. So I don't think this is an artifact of using denoising prior to source extraction. I have not used CNMF-E. Do you still recommend it, after reading this reply? Thanks!

https://github.com/flatironinstitute/CaImAn/assets/20496060/5f24a820-75e7-4c7a-9269-3f6611847c2d
<img width="683" alt="Screen Shot 2024-06-19 at 12 06 03 AM" src="https://github.com/flatironinstitute/CaImAn/assets/20496060/fb641b99-9f1d-4fea-aaf2-fb8814b2ce5f">
<img width="1512" alt="Screen Shot 2024-06-18 at 11 53 31 PM" src="https://github.com/flatironinstitute/CaImAn/assets/20496060/07219535-d586-4a16-b3a4-68985a2f4229">

**Comment by wienecke** [comment]
Hi. Do you still recommend CNMFE after reading my reply? Why does bas_nonneg=False have no effect when p=0? Is this an 'issue'? Should I post this to the 'issues' thread? Thanks

**Comment by kushalkolar** [comment]
A lot of this is _try and see what it looks like_, I've never worked with movies like this so I cannot say for sure, you just have to try and see what happens.

**Comment by wienecke** [comment]
Thanks for your suggestion. Following the demo, I made these changes to run CNMFE:

        method_init = 'corr_pnr' 
        nb = 0 
        only_init = True 
        low_rank_background = None 
        center_psf = True  
        normalize_init = False

Otherwise my parameters were the same as before. 

I still see the clipping of negative values (see screenshot). That is, bas_nonneg=False has no effect when p=0. 

I do not have the math background to understand why the code was written this way. This seems like a potential problem for anybody working with nonspiking neurons. Or maybe the negative values have a meaning that I don't understand. 

Do you know why bas_nonneg=False has no effect when p=0? Do you think this should be posted to 'Issues'? 

Thanks!
Carl

<img width="1069" alt="Screen Shot 2024-08-15 at 11 04 29 PM" src="https://github.com/user-attachments/assets/39daf5fe-15ce-453d-afb3-a60bd963bf3a">

**Comment by kushalkolar** [comment]
Sorry I never dug that much into CNMFE, we currently don't have someone very experienced with CNMFE. What if you disable deconvolution (but I think it should be disabled with CNMFE anyways)
